lib/msf/core/exploit/browser_autopwnv2.rb

```diff
@@ -0,0 +1,834 @@
+    #
+    # @see #bap_exploits
+    # @return [Hash] A module list grouped by rank.
+    def group_bap_modules
+      bap_groups = {}
+      RankingName.each_pair do |ranking, value|
+        bap_groups[ranking] = []
+        bap_exploits.each do |m|
+          next if m.rank != ranking
+          bap_groups[ranking] << m
+        end
+      end
+      bap_groups
+    end
+
+
+    # Modifies @bap_exploit by replacing it with the rearranged module list.
+    #
+    # @see #bap_exploits The read-only attribute.
+    # @param bap_groups [Hash] A grouped module list.
+    # @return [void]
+    def finalize_sorted_modules(bap_groups)
+      @bap_exploits = []
+      bap_groups.each_pair do |ranking, module_list|
+        module_list.each do |m|
+          break if @bap_exploits.length >= datastore['MaxExploits']
+          @bap_exploits << m
+        end
+      end
+    end
+
+
+    # Returns a payload name. Either this will be the user's choice, or falls back to a default one.
+    #
+    # @see DEFAULT_PAYLOADS The default settings.
+    # @param platform [Symbol] Platform name.
+    # @return [String] Payload name.
+    def get_selected_payload_name(platform)
+      payload_name = datastore["PAYLOAD_#{platform.to_s.upcase}"]
+
+      # The payload is legit, we can use it.
+      # Avoid #create seems faster
+      return payload_name if framework.payloads.keys.include?(payload_name)
+
+      default = DEFAULT_PAYLOADS[platform][:payload]
+
+      # The user has configured some unknown payload that we can't use,
+      # fall back to default.
+      default
+    end
+
+
+    # Returns the selected payload's LPORT.
+    #
+    # @param platform [Symbol]
+    # @return [Fixnum]
+    def get_selected_payload_lport(platform)
+      datastore["PAYLOAD_#{platform.to_s.upcase}_LPORT"]
+    end
+
+
+    # Returns the selected payload's LHOST. If no LHOST is set by the user (via the datastore option),
+    # then the method automatically generates one by Rex.
+    #
+    # @return [String]
+    def get_payload_lhost
+      datastore['LHOST'] || Rex::Socket.source_address
+    end
+
+
+    # Creates payload listeners. The active job IDs will be tracked in #payload_job_ids so that
+    # we know how to find them and then clean them up.
+    #
+    # @note FireFox payload is skipped because there's no handler for it.
+    # @see #payload_job_ids
+    # @return [void]
+    def start_payload_listeners
+      # Spawn nothing if the user doesn't want to pop sessions.
+      return if datastore['MaxSessions'] == 0
+
+      # Don't repeat launching payload handlers
+      wanted_payloads.uniq! { |e| e[:payload_name] }
+
+      wanted_payloads.each do |wanted|
+        multi_handler = framework.exploits.create('multi/handler')
+
+        # We have to special case firefox
+        payload_name = wanted[:payload_name].include?('firefox/') ? wanted[:payload_name].gsub('firefox/', 'generic/') : wanted[:payload_name]
+
+        # User configurable options
```

Comment:
Other options that should get passed down should include:
 * HandlerSSLCert
 * StagerVerifySSLCert
 * PayloadUUIDTracking
 * PayloadUUIDName
 * SessionRetryTotal
 * SessionRetryWait
 * SessionExpirationTimeout
 * SessionCommunicationTimeout